math/src/main/java/smile/math/matrix/SparseMatrix.java

```diff
@@ -175,11 +180,204 @@ public int size() {
+                for (; index < m.colIndex[col + 1]; index++) {
+                    action.accept(new SparseMatrixEntry(m.rowIndex[index], col, m.x[index], index, m.x));
+                }
+            }
+        }
+
+        public boolean tryAdvance(Consumer<? super SparseMatrixEntry> action) {
+            if (col < fence) {
+                while (col < fence && index >= m.colIndex[col + 1]) {
+                    col++;
+                }
+                if (col < fence) {
+                    action.accept(new SparseMatrixEntry(m.rowIndex[index], col, m.x[index], index, m.x));
+                    index++;
+                    return true;
+                } else {
+                    return false;
+                }
+            } else {
+                // cannot advance
+                return false;
+            }
+        }
+
+        public Spliterator<SparseMatrixEntry> trySplit() {
+            int lo = col; // divide range in half, assuming roughly constant size of all columns
+            int mid = ((lo + fence) >>> 1) & ~1; // force midpoint to be even
+            if (lo < mid) { // split out left half
+                col = mid; // reset this Spliterator's origin
+                return new SparseMatrixSpliterator(m, lo, mid);
+            } else {
+                // too small to split
+                return null;
+            }
+        }
+
+        public long estimateSize() {
+            return (long) (m.colIndex[fence] - m.colIndex[col]);
+        }
+
+        public int characteristics() {
+            return ORDERED | SIZED | IMMUTABLE | SUBSIZED;
+        }
+    }
+
+    /**
+     * Encapsulates important information about an entry in a matrix for use
+     * in streaming, including a string that leads back to the original cell
+     * so that in-place updates are possible.
+     */
+    public static class SparseMatrixEntry {
+        // these fields are exposed for direct access to simplify in-lining by the JVM
+        public final int row;
+        public final int col;
+        public double x;
+
+        // these are hidden due to internal dependency
+        private final int index;
+        private double[] values;
+
+
+        @SuppressWarnings("WeakerAccess")
+        public SparseMatrixEntry(int row, int col, double x, int index, double[] values) {
+            this.row = row;
+            this.col = col;
+            this.x = x;
+            this.index = index;
+            this.values = values;
+        }
+
+        public void set(double value) {
+            this.x = value;
+            values[index] = value;
+        }
+
+        public SparseMatrixEntry clone(double value) {
+            return new SparseMatrixEntry(row, col, value, index, values);
+        }
+    }
+
+    /**
+     * Functional interface for lambda iteration through a sparse matrix.
+     */
+    public interface MatrixElementConsumer {
+        void eval(int row, int column, double value);
```

Comment:
`accept` is typically used as the method name of a consumer.

Reply:
Fine.

The name almost never appears in user code anyway.

A consumer, however, normally only takes a single argument, I think.

Perhaps a better approach is to to change the name from Consumer to Action or Function.

Reply:
`Function` returns values while `Consumer` not. `Action` is mostly used in GUI things. There are `BiConsumer` taking two arguments. I think that it is fine to call it `Consumer`. `accept` is better name than `eval`. We are not really evaluate the element value here.

Reply:
Will switch to accept.

Reply:
Done.